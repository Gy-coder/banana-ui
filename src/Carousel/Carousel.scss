.g-carousel {
    &-window {
      overflow: hidden;
      position: relative;
    }
    &-wrapper {
      position: relative;
    }
    &-dots {
      display: flex;
      justify-content: center;
      align-items: center;
      flex-direction: row;
      position: absolute;
      left: 50%;
      transform: translateX(-50%);
      bottom: 10%;
      z-index: 15;
      > span {
        position: relative;
        display: inline-block;
        flex: 0 1 auto;
        box-sizing: content-box;
        width: 24px;
        height: 4px;
        margin: 0 6px;
        padding: 0;
        text-align: center;
        text-indent: -999px;
        vertical-align: top;
        transition: all 0.5s;
        > button {
          display: block;
          width: 100%;
          height: 4px;
          padding: 0;
          color: transparent;
          font-size: 0;
          background: #fff;
          border: 0;
          border-radius: 1px;
          outline: none;
          cursor: pointer;
          opacity: 0.3;
          transition: all 0.5s;
          &.active {
            background: #fff;
            opacity: 1;
          }
        }
        &:hover {
          cursor: pointer;
        }
        &.active {
          width: 30px;
          background: white;
        }
      }
    }
    &-button {
      z-index: 15;
      position: absolute;
      top: 50%;
      width: 32px;
      height: 48px;
      background: rgba(#000, 0.3);
      border-radius: 20px;
      cursor: pointer;
      color: white;
      display: flex;
      justify-content: center;
      align-items: center;
      &.g-carousel-button-left {
        left: 0;
        transform: translate(-25%, -50%);
      }
      &.g-carousel-button-right {
        right: 0;
        transform: translate(25%, -50%);
      }
    }
  }
  
  .g-carousel-item {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: -1;
    visibility: hidden;
    &.active {
      position: relative;
      z-index: 1;
      visibility: visible;
    }
  }
  
  .g-carousel-item.slide {
    &.carousel-enter {
      transform: translateX(100%);
      visibility: visible;
      transition: none;
      &.reverse {
        transform: translateX(-100%);
      }
    }
    &.carousel-enter-active {
      transition: all 0.5s ease;
      transform: translateX(0) !important;
      visibility: visible;
    }
    &.carousel-enter-done {
      visibility: visible;
    }
  
    &.carousel-exit {
      transform: translateX(0);
      visibility: visible;
    }
    &.carousel-exit-active {
      transition: all 0.5s ease;
      transform: translateX(-100%);
      visibility: visible;
      &.reverse {
        transform: translateX(100%);
      }
    }
    &.carousel-exit-done {
      visibility: hidden;
    }
  }
  