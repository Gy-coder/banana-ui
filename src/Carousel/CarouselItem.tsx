import React, { CSSProperties, FC, Fragment, useMemo } from 'react';
import { CSSTransition } from 'react-transition-group';
import classnames from 'classnames';

export interface ItemProps {
  curIndex?: number;
  index?: number;
  reverse?: boolean;
  style?: CSSProperties;
  animationName?: string;
}

const carouselItem: FC<ItemProps> = (props) => {
  const { children, index, curIndex, reverse, animationName = 'slide' } = props;
  const active = useMemo(() => curIndex === index, [curIndex]);
  return (
    <Fragment>
      <CSSTransition
        in={active}
        classNames="carousel"
        timeout={3000}
        key={index}
        // unmountOnExit
      >
        <div
          className={classnames('g-carousel-item', animationName, {
            active,
            reverse,
          })}
        >
          {children}
        </div>
      </CSSTransition>
    </Fragment>
  );
};

export default carouselItem;
