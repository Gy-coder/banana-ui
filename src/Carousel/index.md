---
title: Carousel
nav:
  title: Carousel
  path: /components
---

# Carousel

旋转木马，一组轮播的区域。

# 何时使用

- 当有一组平级的内容。
- 当内容空间不足时，可以用走马灯的形式进行收纳，进行轮播展现。
- 常用于一组图片或卡片轮播。

# 代码演示

## 基础使用

<code src="./demos/basic.tsx" />

## 不自动播放

<code src="./demos/notAutoPlay.tsx" />

## API

| 参数      | 说明               | 类型          | 是否必须 | 默认值 |
| --------- | ------------------ | ------------- | -------- | ------ |
| time      | 轮播切换的时间     | number        | false    | 3      |
| autoPlay  | 是否自动轮播       | boolean       | false    | true   |
| className | 轮播外层容器的类名 | string        | false    |        |
| style     | 轮播外层容器的样式 | CSSProperties | false    |        |
