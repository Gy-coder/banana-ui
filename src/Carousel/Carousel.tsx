import React, {
  CSSProperties,
  FC,
  ReactNode,
  TouchEventHandler,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import CarouselItem, { ItemProps } from './CarouselItem';
import classnames from 'classnames';
import { LeftOutlined, RightOutlined } from '@ant-design/icons';
import './Carousel.scss';

export interface Props {
  time?: number;
  autoPlay?: boolean;
  className?: string;
  style?: CSSProperties;
}
interface Params {
  curIndex: number;
  direction: 'forward' | 'reverse';
}

const Component: FC<Props> = (props) => {
  const { children, time = 3, autoPlay, className, style } = props;
  const [params, setParams] = useState<Params>({
    curIndex: 0,
    direction: 'forward',
  });
  const timerId = useRef<any>(null);
  const length = useMemo(() => {
    return Array.from(children as ReactNode[]).length || 1;
  }, [children]);
  const [showButton, setShowButton] = useState(false);
  const classes = classnames('g-carousel', className);
  const startTouch = useRef<any>();
  const endTouch = useRef<any>();
  const onNext = () => {
    setParams(() => {
      params.curIndex = (params.curIndex + 1 + length) % length;
      params.direction = 'forward';
      return { ...params };
    });
  };
  const onPrev = () => {
    setParams(() => {
      params.curIndex = (params.curIndex - 1 + length) % length;
      params.direction = 'reverse';
      return { ...params };
    });
  };
  const playAutomatically = () => {
    if (!autoPlay) return;
    timerId.current = window.setInterval(onNext, time * 1000);
  };
  const parse = () => {
    window.clearInterval(timerId.current);
  };
  const handleMouseEnter = () => {
    parse();
    setShowButton(true);
  };
  const handleMouseLeave = () => {
    playAutomatically();
    setShowButton(false);
  };
  const handleClickDot = (idx: number) => {
    setParams(() => {
      params.direction = params.curIndex > idx ? 'reverse' : 'forward';
      params.curIndex = idx;
      return { ...params };
    });
  };
  const onTouchStart: TouchEventHandler<HTMLDivElement> = (e) => {
    parse();
    if (e.touches.length > 1) return;
    startTouch.current = e.touches[0];
  };
  const onTouchEnd: TouchEventHandler<HTMLDivElement> = (e) => {
    endTouch.current = e.changedTouches[0];
    const { clientX: x1, clientY: y1 } = startTouch.current;
    const { clientX: x2, clientY: y2 } = endTouch.current;
    const distance = Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
    const deltaY = Math.abs(y2 - y1);
    const rate = distance / deltaY;
    if (rate > 2) {
      if (x2 > x1) {
        onNext();
      } else {
        onPrev();
      }
    }
    if (props.autoPlay) playAutomatically();
  };
  useEffect(() => {
    playAutomatically();
    return () => clearInterval(timerId.current);
  }, []);
  return (
    <div
      className={classes}
      style={style}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      onTouchStart={onTouchStart}
      onTouchEnd={onTouchEnd}
    >
      <div className="g-carousel-window">
        <div className="g-carousel-wrapper">
          {React.Children.map(children, (child, index) => {
            const ChildElement =
              child as React.FunctionComponentElement<ItemProps>;
            //@ts-ignore
            if (child.type !== CarouselItem) throw new Error('必须是Item');
            return React.cloneElement(ChildElement, {
              index,
              curIndex: params.curIndex,
              reverse: params.direction === 'reverse',
            });
          })}
        </div>
        <div className="g-carousel-dots">
          {Array.from({ length }).map((item, index) => {
            return (
              <span
                className={classnames('g-carousel-dot', {
                  active: params.curIndex === index,
                })}
                onClick={() => handleClickDot(index)}
                key={item as number}
              >
                <button />
              </span>
            );
          })}
        </div>
        <div
          className="g-carousel-button g-carousel-button-left"
          style={{ display: !showButton ? 'none' : 'flex' }}
          onClick={onPrev}
        >
          <span>
            <LeftOutlined />
          </span>
        </div>
        <div
          className="g-carousel-button g-carousel-button-right"
          style={{ display: !showButton ? 'none' : 'flex' }}
          onClick={onNext}
        >
          <span>
            <RightOutlined />
          </span>
        </div>
      </div>
    </div>
  );
};

type CarouselType = {
  Item: FC<ItemProps>;
} & FC<Props>;

const Carousel: CarouselType = Component as CarouselType;

Carousel.Item = CarouselItem;

Carousel.defaultProps = {
  time: 3,
  autoPlay: true,
};

export default Carousel;
