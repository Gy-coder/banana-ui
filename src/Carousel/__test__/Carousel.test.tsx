import React from 'react';
import {
  render,
  screen,
  RenderResult,
  fireEvent,
  cleanup,
  act,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Carousel from '../Carousel';

import { config } from 'react-transition-group';

config.disabled = true;

describe('test Carousel Component', () => {
  let wrapper: RenderResult;
  beforeEach(() => {
    wrapper = render(
      <Carousel>
        <Carousel.Item>
          <div className="box box1">0</div>
        </Carousel.Item>
        <Carousel.Item style={{ background: '#fff' }}>
          <div className="box box1">1</div>
        </Carousel.Item>
        <Carousel.Item>
          <div className="box box1">2</div>
        </Carousel.Item>
      </Carousel>,
    );
  });
  it('Carousel can correctly render', () => {
    expect(screen.getByText('0')).toBeInTheDocument();
    expect(wrapper.asFragment()).toMatchSnapshot();
    expect(screen.getByText('0').parentNode).toHaveClass('active');
  });
  it('Carousel should click to next and prev', () => {
    const prevButton = wrapper.container.querySelector(
      '.g-carousel-button-left',
    );
    const nextButton = wrapper.container.querySelector(
      '.g-carousel-button-right',
    );
    expect(prevButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
    fireEvent.click(nextButton);
    expect(screen.getByText('0').parentNode).not.toHaveClass('active');
    expect(screen.getByText('1').parentNode).toHaveClass('active');
    expect(screen.getByText('2').parentNode).not.toHaveClass('active');
    fireEvent.click(nextButton);
    expect(screen.getByText('0').parentNode).not.toHaveClass('active');
    expect(screen.getByText('1').parentNode).not.toHaveClass('active');
    expect(screen.getByText('2').parentNode).toHaveClass('active');
    fireEvent.click(prevButton);
    expect(screen.getByText('0').parentNode).not.toHaveClass('active');
    expect(screen.getByText('1').parentNode).toHaveClass('active');
    expect(screen.getByText('2').parentNode).not.toHaveClass('active');
  });
  it('click dot can change page', () => {
    cleanup();
    wrapper = render(
      <Carousel>
        <Carousel.Item>
          <div className="box box1">0</div>
        </Carousel.Item>
        <Carousel.Item style={{ background: '#fff' }}>
          <div className="box box1">1</div>
        </Carousel.Item>
        <Carousel.Item>
          <div className="box box1">2</div>
        </Carousel.Item>
        <Carousel.Item>
          <div className="box box1">3</div>
        </Carousel.Item>
        <Carousel.Item style={{ background: '#fff' }}>
          <div className="box box1">4</div>
        </Carousel.Item>
        <Carousel.Item>
          <div className="box box1">5</div>
        </Carousel.Item>
      </Carousel>,
    );
    const dots = wrapper.container.querySelectorAll('.g-carousel-dot');
    expect(dots.length).toEqual(6);
    expect(dots[0]).toHaveClass('active');
    for (let i = 1; i < dots.length; i++) {
      expect(dots[i]).not.toHaveClass('active');
    }
    fireEvent.click(dots[3]);
    expect(dots[3]).toHaveClass('active');
    expect(dots[0]).not.toHaveClass('active');
  });
  it('test auto play', () => {
    jest.useFakeTimers();
    cleanup();
    wrapper = render(
      <Carousel autoPlay>
        <Carousel.Item>
          <div className="box box1">0</div>
        </Carousel.Item>
        <Carousel.Item style={{ background: '#fff' }}>
          <div className="box box1">1</div>
        </Carousel.Item>
        <Carousel.Item>
          <div className="box box1">2</div>
        </Carousel.Item>
      </Carousel>,
    );
    expect(screen.getByText('0').parentNode).toHaveClass('active');
    act(() => {
      jest.advanceTimersByTime(3000);
    });
    expect(screen.getByText('0').parentNode).not.toHaveClass('active');
    expect(screen.getByText('1').parentNode).toHaveClass('active');
    jest.clearAllTimers();
  });
  it('can set timeout', () => {
    jest.useFakeTimers();
    cleanup();
    wrapper = render(
      <Carousel autoPlay time={5}>
        <Carousel.Item>
          <div className="box box1">0</div>
        </Carousel.Item>
        <Carousel.Item style={{ background: '#fff' }}>
          <div className="box box1">1</div>
        </Carousel.Item>
        <Carousel.Item>
          <div className="box box1">2</div>
        </Carousel.Item>
      </Carousel>,
    );
    expect(screen.getByText('0').parentNode).toHaveClass('active');
    act(() => {
      jest.advanceTimersByTime(5000);
    });
    expect(screen.getByText('0').parentNode).not.toHaveClass('active');
    expect(screen.getByText('1').parentNode).toHaveClass('active');
    act(() => {
      jest.advanceTimersByTime(5000);
    });
    expect(screen.getByText('1').parentNode).not.toHaveClass('active');
    expect(screen.getByText('2').parentNode).toHaveClass('active');
    jest.clearAllTimers();
  });
  xit('test touch event', () => {
    const outline = wrapper.container.querySelector('.g-carousel');
  });
});
