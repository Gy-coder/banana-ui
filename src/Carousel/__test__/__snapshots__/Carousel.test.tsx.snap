// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Carousel Component Carousel can correctly render 1`] = `
<DocumentFragment>
  <div
    class="g-carousel"
  >
    <div
      class="g-carousel-window"
    >
      <div
        class="g-carousel-wrapper"
      >
        <div
          class="g-carousel-item slide active"
        >
          <div
            class="box box1"
          >
            0
          </div>
        </div>
        <div
          class="g-carousel-item slide"
        >
          <div
            class="box box1"
          >
            1
          </div>
        </div>
        <div
          class="g-carousel-item slide"
        >
          <div
            class="box box1"
          >
            2
          </div>
        </div>
      </div>
      <div
        class="g-carousel-dots"
      >
        <span
          class="g-carousel-dot active"
        >
          <button />
        </span>
        <span
          class="g-carousel-dot"
        >
          <button />
        </span>
        <span
          class="g-carousel-dot"
        >
          <button />
        </span>
      </div>
      <div
        class="g-carousel-button g-carousel-button-left"
        style="display: none;"
      >
        <span>
          <span
            aria-label="left"
            class="anticon anticon-left"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="left"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="g-carousel-button g-carousel-button-right"
        style="display: none;"
      >
        <span>
          <span
            aria-label="right"
            class="anticon anticon-right"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
